astropy/coordinates/builtin_frames/astrometric.py

```diff
@@ -137,16 +136,57 @@ def icrs_to_astrometric(reference_frame, astrometric_frame):
         R = mat1 * mat2 * mat3
         return R
 
-    @frame_transform_graph.transform(DynamicMatrixTransform, Astrometric, framecls)
+    @frame_transform_graph.transform(DynamicMatrixTransform, _Astrometric, framecls)
     def astrometric_to_icrs(astrometric_coord, reference_frame):
-        """Convert an Astrometric frame coordinate to an ICRS"""
+        """Convert an Astrometric frame coordinate to the reference frame"""
 
         # use the forward transform, but just invert it
         R = icrs_to_astrometric(reference_frame, astrometric_coord)
         return R.T  # this is the inverse because R is a rotation matrix
 
-    _astrometric_cache[framecls] = Astrometric
-    return Astrometric
+    _astrometric_cache[framecls] = _Astrometric
+    return _Astrometric
+
+class AstrometricFrame(BaseCoordinateFrame):
+    """
+    A frame which is relative to some position on the sky. Useful for
+    calculating offsets and dithers in the frame of the sky.
+
+    Parameters
+    ----------
+    representation : `BaseRepresentation` or None
+        A representation object or None to have no data (or use the other keywords)
+    origin : `SkyCoord` or low-level coordinate object.
+        the coordinate which specifiy the origin of this frame.
+    rotation : Quantity with angle units
+        The final rotation of the frame about the ``origin``. The sign of
+        the rotation is the left-hand rule.  That is, an object at a
+        particular position angle in the un-rotated system will be sent to
+        the positive latitude (z) direction in the final frame.
+
+    """
 
-AstrometricICRS = make_astrometric_cls(ICRS)
-AstrometricAltAz = make_astrometric_cls(AltAz)
+    rotation = QuantityFrameAttribute(default=0, unit=u.deg)
+    origin = CoordinateAttribute(default=None, frame=None)
+
+    def __new__(cls, *args, **kwargs):
+        # We don't want to call this method if we've already set up
+        # an astrometric frame for this class.
+        if not (issubclass(cls, AstrometricFrame) and cls is not AstrometricFrame):
+            # We get the origin argument, and handle it here.
+            try:
+                origin_frame = kwargs['origin']
+            except KeyError:
+                raise TypeError("Can't initialize an AstrometricFrame without origin= keyword.")
+            if hasattr(origin_frame, 'frame'):
+                origin_frame = origin_frame.frame
+            newcls = make_astrometric_cls(origin_frame.__class__)
+            return newcls.__new__(newcls, *args, **kwargs)
+
+        # http://stackoverflow.com/questions/19277399/why-does-object-new-work-differently-in-these-three-cases
+        # See above for why this is necessary. Basically, because some child
+        # may override __new__, we must override it here to never pass
+        # arguments to the object.__new__ method.
+        if super(AstrometricFrame, cls).__new__ is object.__new__:
```

Comment:
In fact, because BaseCoordinateFrame doesn't implement ``__new__``, this is always triggered right now. If another frame were to implement a custom ``__new__``, this would be triggered by the parent class when traversing the class hierarchy if ``AstrometricFrame`` came before the ``framecls`` above, which ideally it would, but that breaks something about ordered descriptor that I don't understand (but maybe @eteq could fix?)

Regardless, right now, this is required to prevent this ``__new__`` method from breaking should an descendant implement ``__new__`` in the future. Since its a subtle bug when that happens, I would advocate for leaving the solution in now.